### DISCUSSION TITLE
Architecture variant: no topics, Web Push (including UnifiedPush) + others

### DISCUSSION THREAD

**Discussion by rfc2822** [bitfireAT/webdav-push#21] [General]
This is our current favorite architecture. Other architectures we have thought about: #16, #17, #18, #20

Update 25 Dec 2023: UnifiedPush is a 100% compatible subset of Web Push from a WebDAV-Push point of view

---

A key decision is whether to support **topics** (= multiple clients subscribe to the same topic and then the application server pushes only one "topic update" and leaves sending the single notifications to the push transport) or not. In this architecture we don't have topics. This allows to keep things simple and, more important, to directly support UnifiedPush. The WebDAV server and especially the FCM redirector have to handle a bit more traffic (one notification per client instead of one topic update for all clients), but this comes with the advantage that software components can be less complex: the redirector doesn't need to keep any state and is thus much easier to build, maintain and host. As soon as we drop the "topics" requirement (which came up when studying proprietary APIs like FCM, who always provide topics and it seems common to use them here), it becomes much easier to use other open standards/components.

We aim to support the possibility of other push transports, but for our primary use cases and planned implementations we focus on

## **[Web Push](https://www.rfc-editor.org/rfc/rfc8030)**

  - for browser apps (like the Nextcloud Calendar app)
  - for desktop and mobile apps over **[UnifiedPush](https://unifiedpush.org/)** (UP)
  - already available in all major Web browsers and for platforms on which UP is supported
  - Web Push supports [VAPID](https://www.rfc-editor.org/rfc/rfc8292) and [Message Encryption](https://datatracker.ietf.org/doc/html/rfc8291) and we want to use it

**UnifiedPush** can be used for mobile (like DAVx⁵) and desktop apps (like desktop Groupware applications or file managers). From an RFC 8030 Application Server point of view, UP endpoints are designed to be 100% compatible RFC 8030 push resources and UP servers can be treated like RFC 8030 Push Services.

  - It's already working and established, [FCM redirector already exists](https://unifiedpush.org/developers/embedded_fcm/#fcm-rewrite-proxy) (when it's even needed [[1]](https://github.com/bitfireAT/webdav-push/discussions/21#discussioncomment-7934652) [[2]](https://github.com/bitfireAT/webdav-push/discussions/21#discussioncomment-7943873), which should be clarified)
  - Further proprietary transports can be implemented over redirect proxies, if required
  - For desktop applications, disadvantage in comparision to [SSE](https://html.spec.whatwg.org/multipage/server-sent-events.html#server-sent-events) directly provided by the Push-capable WebDAV server is that apps need a UnifiedPush distributor / server. However for instance [KDE has attempts in this direction](https://www.volkerkrause.eu/2022/11/12/kde-unifiedpush-push-notifications.html). It has more chances to add an existing protocol (UnifiedPush) and required infrastructure to desktop clients instead of defining and adopting a new protocol (WebDAV-Push over SSE).
  - E2EE can be implemented on application level: https://github.com/UnifiedPush/specifications/issues/12

![Architecture](https://raw.githubusercontent.com/bitfireAT/webdav-push/5adf1c80695e9c3c77c302278738485862acb772/images/architecture.png)

**Comment by mar-v-in** [comment]
Firebase Cloud Messaging is directly compatible to WebPush notifications. This is used by Chrome and some other Chromium-based browsers to receive WebPush notifications through the Firebase Cloud Messaging infrastructure (which is also used by Chrome). I wrote down how to register as a client app with FCM to receive a token that can be used to push using WebPush protocol here: https://gist.github.com/mar-v-in/2a054e3a4c0a508656549fc7d0aaeb74#webpush This would allow to completely get rid of the redirect proxy when sending to FCM.

You can also send WebPush notifications to a UnifiedPush endpoint. UnifiedPush currently implements a subset of WebPush (i.e. it does not use VAPID), but it still would delivers the push notification to the client. I think it would make more sense to extend UnifiedPush server-side protocol to fully support the WebPush protocol rather than supporting WebPush AND UnifiedPush in every protocol. UnifiedPush would still be responseible for delivering the push notification to the client on the client-side (using the APIs specified in the Android and DBus specification documents of the UnifiedPush protocol).

Limiting this whole specification to use WebPush exclusively reduces complexity in the specification. If a vendor wants to use any other protocols than WebPush, their best way to do so is to provide a WebPush redirect proxy. Let's unify on one common standard for the server side of push notifications - and with WebPush already having wide support (on Google Android and all major browsers) and being an IETF standard, it seems the obvious choice for that.

**Reply by karmanyaahm** [reply]
Absolutely agree with @mar-v-in. The UnifiedPush team intentionally made it support the same core protocol as RFC8030 on the push side so it can carry WebPush messages. Only supporting WebPush will make implementation on the server side far easier, and all UnifiedPush clients will still work with it. Here are some great implementations of RFC8291 decryption in UnfiedPush apps: [lttrs-android](https://codeberg.org/iNPUTmice/lttrs-android/src/commit/c0ac0ec214a2f4dbf1eb27b90af720434dc15f79/app/src/main/java/rs/ltt/android/push/WebPushMessageEncryption.java) and [Podverse](https://github.com/podverse/podverse-rn/blob/develop/android/app/src/main/java/com/podverse/PVUnifiedPushEncryption.java)

**Reply by rfc2822** [reply]
Well even better... what I find strange is that Web Push is not mentioned with a single word on the Unified Push homepage? Or did I just overlook it? For me it seemed like the coincidence was rather by "chosen randomness" and that one shouldn't count on it in the details.

Also, I dont think it will make the server side much simpler. In any case it would just send a POST, so it shouldn't be that much difference? And in any case we want to support multiple transports. Maybe Web Push (or UnifiedPush) is the rising star today, but nobody knows about tomorrow and maybe other platforms. So the protocol should at least provide the possibility for other transports.

So you suggest to drop the UnifiedPush on server side and send Web Push with VAPID and Message Encryption directly to the UnifiedPush endpoints, as if they were Web Push push resources?

What about the things that are specified in Web Push, like, lets say Replacing messages with the topic header? Can one count on UnifiedPush to support these things? If not, where can I find a table of what is supported what not? I must confess that this "we have two protocols which are not the same, but look the same, work similarly and aim to be exchangable but its not mentioned anywhere" thing confuses me very much. But maybe I have overlooked something and you can point me to more information, which would be great.

**Reply by mar-v-in** [reply]
UnifiedPush tries to intentionally be compatible with WebPush. SERV_1.2.0 specification added the requirement to add a `TTL: 0` header to the response. This makes the server technically WebPush compatible without supporting topic-based message replacement. Additionally there is discussion to support other WebPush features in the future, see https://github.com/UnifiedPush/specifications/pull/37

In any case, as an application server, you're supposed to expect that a UnifiedPush push server behaves WebPush compatible. The `TTL: 0` header is 100% valid on WebPush as well (meaning any WebPush server could do this and you need to be prepared for it). If you wanted to discover if a WebPush server supports replacing by topic, you could inspect the `TTL` header, however you have no guarantee whatsoever that a regular WebPush endpoint supports all features of the WebPush protocol even if they announce a `TTL` greater than 0.

**Reply by mar-v-in** [reply]
> So the protocol should at least provide the possibility for other transports.

I don't necessarily agree with that.
- You can always redirect WebPush to something else using a push redirector service if needed. In general, the redirector service infrastructure is probably very sensible as it means that WebDAV server implementations don't need to support more than one push protocol themselves. Redirectors can be shared across completely independent WebDAV server implementations.
- If WebPush was to phase out eventually, creating a new version of this specification "WebDAV Push 2" is certainly possible.

That said, it certainly also doesn't do a lot of damage to have an extension point in this version of the protocol where a future specification (or vendor specific extensions) can tie in, similar to how it currently is.
```xml
<multistatus xmlns="DAV:">
  <response>
    <href>/webdav/collection/</href>
    <prop>
      <push-transports xmlns="DAV:Push">
        <web-push>
          <public-key>...</public-key>
        </web-push>
        <!-- Additional vendor-specific or future standard options can go here using their own namespace, e.g. -->
        <apns xmlns="https://apple.com/webdav/push" />
      </push-transports>
    </prop>
  </response>
</multistatus>
```

**Reply by karmanyaahm** [reply]
> Well even better... what I find strange is that Web Push is not mentioned with a single word on the Unified Push homepage? Or did I just overlook it? For me it seemed like the coincidence was rather by "chosen randomness" and that one shouldn't count on it in the details.

Good point, that's my bad. Adding "app server implementation notes on UP.org" on my to-do list. So far most implementations have just been discussions in the UP matrix rooms, which I agree is not great from a documentation standpoint.

> So you suggest to drop the UnifiedPush on server side and send Web Push with VAPID and Message Encryption directly to the UnifiedPush endpoints, as if they were Web Push push resources?

As the person who wrote the original UP server protocol specification, I can say that this is absolutely correct. And, this is the direction we intend to move the spec in. 

Regardless of how this ends up being implemented I'm really looking forward to receiving instant calendar updates from my Nextcloud server on my phone! hopefully with E2EE :)

**Reply by rfc2822** [reply]
Ok, thanks for the info, I will adapt things accordingly!

**Reply by rfc2822** [reply]
> Firebase Cloud Messaging is directly compatible to WebPush notifications. This is used by Chrome and some other Chromium-based browsers to receive WebPush notifications through the Firebase Cloud Messaging infrastructure (which is also used by Chrome). I wrote down how to register as a client app with FCM to receive a token that can be used to push using WebPush protocol here: https://gist.github.com/mar-v-in/2a054e3a4c0a508656549fc7d0aaeb74#webpush This would allow to completely get rid of the redirect proxy when sending to FCM.

That's interesting. I had the same idea and tried to find out the push resource URL from official FCM docs, but didn't succeed.

If we wouldn't have to operate a redirect proxy for DAVx5, this would be very handy. On the other side I wonder whether just specifiying [the FCM URL as you describe it](https://gist.github.com/mar-v-in/2a054e3a4c0a508656549fc7d0aaeb74#sending-messages) as push resource is using undocumented behavior that is subject of change at any time (for instance, they could block non-authenticated push requests when the target is Android) or whether it's officially supported by Google. And if it's not supported officially, whether it wouldn't be better to use the redirect proxy which is the "official" way.

Do you maybe know how it's handled by Huawei (which we also want to support, especially for the Google-free Huawei devices)? They have a [Huawei Browser](https://consumer.huawei.com/en/mobileservices/browser/), they have a [Push Kit](https://developer.huawei.com/consumer/en/hms/huawei-pushkit) and so I guess they support Web Push and will have a similar infrastructure. However the [official docs also say that an application server has to authenticate to send downlink messages](https://developer.huawei.com/consumer/en/doc/HMSCore-Guides/android-server-dev-0000001050040110).

**Reply by mar-v-in** [reply]
> On the other side I wonder whether just specifiying [the FCM URL as you describe it](https://gist.github.com/mar-v-in/2a054e3a4c0a508656549fc7d0aaeb74#sending-messages) as push resource is using undocumented behavior that is subject of change at any time (for instance, they could block non-authenticated push requests when the target is Android) or whether it's officially supported by Google.

I'm not sure in how far this behavior is official (it certainly is not properly documented), but it is used by Chrome and third-party Chromium-based browsers on Android (including Opera and Edge). Some of them have slightly different behavior during the registration (e.g. Edge registers with `edge-webpush:{origin}#{randomUuid}` instead of `wp:{randomUuid}`, other variants include `wp:{origin}#{randomUuid}`, `wp:{origin}#{randomUuid, but with last three chars replaced by '-V2'}`), but for all of them the resulting push endpoint is `https://fcm.googleapis.com/fcm/send/<token>`. What is obviously still required is that the app requesting such push token is registered with Google Cloud Messaging (including its signing certificate). Additionally, Chrome internally uses this endpoint to send push notifications to other Chrome instances when you have multiple Chrome instances signed into the same Google account (see [here](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/sharing/web_push/web_push_sender.cc;l=91;drc=8e78783dc1f7007bad46d657c9f332614e240fd8;bpv=1;bpt=1)).

As third-parties are involved and this endpoint is spread widely (it is stored on the application servers of web apps registering in all those browsers), I highly doubt Google is going to change anything for this endpoint. They might change requirements in the registration procedure, but even that would require probably multiple years of preparation to ensure old browsers are not used anymore.

> Do you maybe know how it's handled by Huawei (which we also want to support, especially for the Google-free Huawei devices)? They have a [Huawei Browser](https://consumer.huawei.com/en/mobileservices/browser/), they have a [Push Kit](https://developer.huawei.com/consumer/en/hms/huawei-pushkit) and so I guess they support Web Push and will have a similar infrastructure. However the [official docs also say that an application server has to authenticate to send downlink messages](https://developer.huawei.com/consumer/en/doc/HMSCore-Guides/android-server-dev-0000001050040110).

I tested Huawei Browser on a Huawei device and it does not support WebPush. Huawei claims in its documentation that PushKit for Web is only supported on Chrome, Firefox and Edge desktop browsers. The Android support of PushKit is only available to native Android apps.

Huawei PushKit can act as a push aggregating proxy: app servers always push to Huawei PushKit servers, and Huawei will distribute the push message to the app using an appropriate channel. The push of the app server to the PushKit server does not use RFC 8030, but to deliver the message through web push, the PushKit server sends it to the browser's designated push server via RFC 8030. It also supports sending via APNS to iOS, FCM to Google Android and HMS to Huawei Android.

**Reply by rfc2822** [reply]
Hm as I understand it, this is at least pseudo-officially alluded in [Web Push Interoperability Wins](https://developer.chrome.com/blog/web-push-interop-wins?hl=en):

> FCM / GCM and Chrome now support the standard [Web Push Protocol](https://tools.ietf.org/html/draft-ietf-webpush-protocol), while sender authentication can be achieved by implementing [VAPID](https://tools.ietf.org/html/draft-thomson-webpush-vapid), meaning your web app no longer needs a 'gcm_sender_id'.
>
>[…]
>
>```
>// Assuming you're not using VAPID (read on), this
>// proprietary header is needed
>if(subscription.endpoint
>    .indexOf('https://android.googleapis.com/gcm/send/') === 0) {
>    headers.append('Authorization', 'GCM_API_KEY');
>}
>```

Also they say the URL is `https://fcm.googleapis.com/fcm/send/ABCD1234` …

---

Regarding Huawei:

> The push of the app server to the PushKit server does not use RFC 8030 […]

That's the interesting point for the WebDAV-Push server, as I understand it. So for Huawei a redirect proxy would still be needed, or does it not require "hard" authentication?

**Reply by rfc2822** [reply]
@mar-v-in I didn't manage to send messages to Android over FCM without sharing a key between the app server (which would be the WebDAV server later) and the client (DAVx5). I have used the Android Firebase API to generate and get a token. This token seems to be bound to the key in `google-service.json` in the app. When I then send a Web Push (with some payload, because we want payload) request to that token

- without authentication, FCM returns 401
- with a random VAPID key pair (request generated by [WebPushDataTestPage](https://mozilla-services.github.io/WebPushDataTestPage/)), FCM says: "The VAPID credentials in the authorization header do not correspond to the credentials used to create the subscriptions."

Do I have to create the Android subscription [like this](https://gist.github.com/mar-v-in/2a054e3a4c0a508656549fc7d0aaeb74#registration-1), with the app server public key (that's also the public VAPID key)? In this case this would not really usable by us, because

1. there may be multiple accounts/servers per client, and I guess we can have only one FCM subscription per client app (because `onNewToken` would indicate the previous token is now invalid)?
2. the registration over a crafted undocumented Intent is, while very interesting to understand how things work, maybe not something that I'd use in a production app …

Or is there another way? How can I test sending a notification over a POST to the device token without having to know the private key that's built into the app? Because that would be the scenario of WebDAV-Push.

**Reply by mar-v-in** [reply]
For reference, the code in Chromium is here: https://source.chromium.org/chromium/chromium/src/+/main:components/gcm_driver/android/java/src/org/chromium/components/gcm_driver/GoogleCloudMessagingV2.java

Also,  directly crafting the corresponding intents without using a library once was officially documented, see e.g. https://web.archive.org/web/20121206054553/http://developer.android.com/google/gcm/gcm.html#registering (of course way less features back then)
I'm not sure if any of the several official libraries to invoke C2DM has the required functionality (subtypes aren't a popular feature, so I'm not sure if they are exposed directly).

**Reply by rfc2822** [reply]
The problem with multiple accounts per client would still persist, no?

Let's say we have WebDAV server A (with key pair A) and WebDAV server B (with key pair B). A client which wants to use FCM (DAVx5) gets only one FCM token, which can only be bound to one public key (A_public or B_public).

Or can multiple tokens be generated and used at the same time? Because the FCM Android API only has a `onNewToken` callback and it looks like if there could only be one active token per app.

Or don't we need multiple tokens and I have understood it incorrect?

(And it may have reasons why they have removed it from the docs?)

**Reply by mar-v-in** [reply]
You get one token per subtype.

**Reply by rfc2822** [reply]
Ok thanks. All these undocumented things sound really dirty… but however it's then a client decision (crafting undocumented Intents vs hosting a redirect proxy) and doesn't matter from a WebDAV-Push point of view :)

**Comment by rfc2822** [comment]
> I don't necessarily agree with that.
> 
>     * You can always redirect WebPush to something else using a push redirector service if needed. In general, the redirector service infrastructure is probably very sensible as it means that WebDAV server implementations don't need to support more than one push protocol themselves. Redirectors can be shared across completely independent WebDAV server implementations.

I wouldn't specify that they have to implement a single backend. May be RFC 8030 but there may be times where RFC 8030 (or a direct successor) isn't used anymore, but WebDAV-Push is. Similar things have happened for many protocols.

> That said, it certainly also doesn't do a lot of damage to have an extension point in this version of the protocol where a future specification (or vendor specific extensions) can tie in, similar to how it currently is.

Where is the problem / "damage" with specifying that multiple transports are supported, and RFC 8030 is currently the (maybe only) recommended one?

Regarding UP / Web Push: I have seen that they are too similar as this could be by chance, however I'm still confused that there's no official reference from one to the other, no compatibility table, no definition about missing features etc. And for instance sending VAPID headers to UP without reason, well yes, one could do it, but it still sounds strange to me.

But I think at the end this question won't matter. It's now time to make these things actually work, and how we call the transport then and whether there are two slightly different ones or one for all, shouldn't really make a difference.

---

My biggest previous questions (without having experience in the "push universe") were whether WebDAV-Push could support topics in the sense of FCM (transport does the 1:n multiplication), and it seems that a 1:1 architecture is a better idea becaues of simplicity, E2EE and existing protocols/components. Another question was whether we can support both open and proprietary (and not only Google FCM, but also Huawei Push and maybe others) push systems, and I'm happy to know that this should work without big extra efforts.

**If someone sees "big" problems with this architecture (like missing components or that something won't work in reality), please tell us here.** In this case we'd had to re-think everything and it should happen as soon as possible. Otherwise we would now build more functional components on all sides (client + WebDAV server) while updating our draft continously.

**Reply by mar-v-in** [reply]
> Where is the problem / "damage" with specifying that multiple transports are supported, and RFC 8030 is currently the (maybe only) recommended one?

The "damage" is complexity, both for the specification itself and implementations.
- If you want to specify the use of UnifiedPush and you want this to eventually turn into an RFC, you would first need to create a proper specification of UnifiedPush server protocol (because, well, proper specifications can only reference other proper specifications and as you noticed yourself, the UnifiedPush server specification is still more like an early draft/sketch than an RFC)
- As you allow multiple transports, you would need to specify how clients should react when they see multiple transports that they support. Register for one or for both? If they registered for UnifiedPush before, but now want to switch to WebPush to make use of encryption, how does that work? What happens if you register the same endpoint for UnifiedPush and WebPush (which is technically possible because UnifiedPush endpoints are also valid WebPush endpoints)? What if the set of supported transports changes on either the server or the client?

> no compatibility table, no definition about missing features etc. And for instance sending VAPID headers to UP without reason, well yes, one could do it, but it still sounds strange to me.

There is also no compatibility table or definition of missing features of the WebPush endpoints of the servers of various browsers. Do you know if the push server used for Firefox on Android verifies VAPID for push notifications? Do you care? As long as push notifications arrive when they are sent to the push server and the push server behaves WebPush compliant in the interaction with the application server, the application server should not really care.

> Otherwise we would now build more functional components on all sides (client + WebDAV server) while updating our draft continously.

Of course this is certainly an option for exploring possibilities and drafting the protocol. I just want to raise awareness for the risk of later unwanted functionalities and complexities ending up in production deployments which might be hard or impossible to reverse and can lead to clients/servers having to be able to handle "legacy systems" for eternity. The smaller the feature set is that you originally plan to support, the less risk you have of this happening. Intentionally not supporting something while leaving it as a future extension point is what makes a specification flexible and future-proof, not trying to think about every possibility in advance.
